File: day03/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
)

const (
	puzzleInput = "input.txt"
)

func main() {
	file, err := os.Open(puzzleInput)
	if err != nil {
		log.Fatal(err)
	}
	defer func() {
		if err := file.Close(); err != nil {
			log.Fatal(err)
		}
	}()

	var nums [][]byte
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		nums = append(nums, []byte(scanner.Text()))
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	if len(nums) == 0 {
		return
	}

	numbits := len(nums[0])

	count := make([]int, numbits)
	for _, i := range nums {
		for n, j := range i {
			if j == '1' {
				count[n]++
			} else {
				count[n]--
			}
		}
	}
	max := make([]byte, numbits)
	min := make([]byte, numbits)
	for n, i := range count {
		if i > 0 {
			max[n] = '1'
			min[n] = '0'
		} else {
			max[n] = '0'
			min[n] = '1'
		}
	}
	fmt.Println("Part 1:", btoi(max)*btoi(min))

	most := nums
	least := nums

	for i := 0; i < numbits; i++ {
		if len(most) < 2 {
			break
		}
		most = findCommon(true, i, most)
	}
	for i := 0; i < numbits; i++ {
		if len(least) < 2 {
			break
		}
		least = findCommon(false, i, least)
	}
	if len(most) == 1 && len(least) == 1 {
		fmt.Println("Part 2:", btoi(most[0])*btoi(least[0]))
	}
}

func findCommon(most bool, pos int, nums [][]byte) [][]byte {
	onecounts := 0
	zerocounts := 0

	for _, i := range nums {
		if i[pos] == '1' {
			onecounts++
		} else {
			zerocounts++
		}
	}

	zeros := zerocounts <= onecounts
	if most {
		zeros = zerocounts > onecounts
	}

	var res [][]byte

	for _, i := range nums {
		if zeros {
			if i[pos] == '0' {
				res = append(res, i)
			}
		} else {
			if i[pos] == '1' {
				res = append(res, i)
			}
		}
	}

	return res
}

func btoi(b []byte) int {
	c := 0
	for _, i := range b {
		k := 0
		if i == '1' {
			k = 1
		}
		c = c<<1 + k
	}
	return c
}

File: day03/src/main.rs
use std::fs::File;
use std::io::prelude::*;
use std::io::BufReader;

const PUZZLEINPUT: &str = "input.txt";

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let file = File::open(PUZZLEINPUT)?;
    let reader = BufReader::new(file);

    let nums = {
        let mut nums = Vec::new();
        for line in reader.lines() {
            nums.push(line?.into_bytes());
        }
        nums
    };

    if nums.len() == 0 {
        return Ok(());
    }

    let numbits = nums[0].len();

    let mut count = vec![0; numbits];
    for i in &nums {
        for (n, &j) in i.iter().enumerate() {
            if j == b'1' {
                count[n] += 1;
            } else {
                count[n] -= 1;
            }
        }
    }
    let mut max = vec![0; numbits];
    let mut min = vec![0; numbits];
    for (n, &i) in count.iter().enumerate() {
        if i > 0 {
            max[n] = b'1';
            min[n] = b'0';
        } else {
            max[n] = b'0';
            min[n] = b'1';
        }
    }
    println!("Part 1: {}", btoi(&max[..]) * btoi(&min[..]));

    let mut most = nums.iter().collect::<Vec<_>>();
    let mut least = nums.iter().collect::<Vec<_>>();

    for i in 0..numbits {
        if most.len() < 2 {
            break;
        }
        most = find_common(true, i, most);
    }
    for i in 0..numbits {
        if least.len() < 2 {
            break;
        }
        least = find_common(false, i, least);
    }

    if most.len() == 1 && least.len() == 1 {
        println!("Part 2: {}", btoi(&most[0][..]) * btoi(&least[0][..]));
    }
    Ok(())
}

fn find_common<'a>(most: bool, pos: usize, nums: Vec<&'a Vec<u8>>) -> Vec<&'a Vec<u8>> {
    let mut onecounts = 0;
    let mut zerocounts = 0;
    for i in &nums {
        if i[pos] == b'1' {
            onecounts += 1;
        } else {
            zerocounts += 1;
        }
    }

    let zeros = if most {
        zerocounts > onecounts
    } else {
        zerocounts <= onecounts
    };

    let mut res = Vec::new();

    for i in nums {
        if zeros {
            if i[pos] == b'0' {
                res.push(i);
            }
        } else {
            if i[pos] == b'1' {
                res.push(i);
            }
        }
    }

    res
}

fn btoi(b: &[u8]) -> i32 {
    let mut c = 0;
    for &i in b {
        c = (c << 1) + (if i == b'1' { 1 } else { 0 });
    }
    return c;
}
